docs/TRICKS.md

```diff
@@ -0,0 +1,290 @@
+
+# Prefer Plutarch level functions
+
+Plutarch level functions have a lot of advantages - they can be hoisted; they are strict so you can [use their arguments however many times you like without duplicating work](#dont-duplicate-work); they are required for Plutarch level higher order functions etc. Unless you _really_ need laziness, like `pif` does, try to use Plutarch level functions.
+
+Also see: [Hoisting](./CONCEPTS.md#hoisting-metaprogramming--and-fundamentals).
+
+# When to use Haskell level functions?
+Although you should generally [prefer Plutarch level functions](#prefer-plutarch-level-functions), there are times when a Haskell level function is actually much better. However, figuring out *when* that is the case - is a delicate art.
+
+There is one simple and straightforward usecase though, when you want a function argument to be lazily evaluated. In such a case, you should use a Haskell level functions that `pdelay`s the argument before calling some Plutarch level function. Recall that [Plutarch level functions are strict](#plutarch-functions-are-strict).
+
+Outside of that straightforward usecase, figuring out when to use Haskell level functions is quite complex. Haskell level functions will always be inlined when generating the Plutus Core. Unless the function is used *only once*, this sort of inlining will increase the script size - which is problematic.
+
+However, if the function is used *only once*, and making it Plutarch level causes extra `plam`s and `#`s to be introduced - you should just make it Haskell level. For example, consider the `pelimList` implementation-
+```hs
+pelimList :: PLift a => Term s (a :--> PBuiltinList a :--> r) -> Term s r -> Term s (PBuiltinList a) -> Term s r
+pelimList match_cons match_nil ls = pmatch ls $ \case
+  PCons x xs -> match_cons # x # xs
+  PNil -> match_nil
+```
+It takes in a Plutarch level function, let's see a typical usage-
+```hs
+pelimList
+  (plam $ \x xs -> pcons # x # (self # xs))
+  pnil
+  ls
+```
+This is rather redundant, the above snippet will exhibit inlining to produce-
+```hs
+pmatch ls $ \case
+  PCons x xs -> (plam $ \x xs -> pcons # x # (self # xs)) # x # xs
+  PNil -> match_nil
+```
+Extra `plam`s and `#`s have been introduced. Really, `pelimList` could have taken a Haskell level function instead-
+```hs
+pelimList :: PLift a => (Term s a -> Term s (PBuiltinList a) :--> Term s r) -> Term s r -> Term s (PBuiltinList a) -> Term s r
+pelimList match_cons match_nil ls = pmatch ls $ \case
+  PCons x xs -> match_cons x xs
+  PNil -> match_nil
+```
+Now, the following usage-
+```hs
+pelimList
+  (\x xs -> pcons # x # (self # xs))
+  pnil
+  ls
+```
+would turn into-
+```hs
+pmatch ls $ \case
+  PCons x xs -> pcons # x # (self # xs)
+  PNil -> match_nil
+```
+It turns out that `pelimList` usages *almost always* use a one-off Haskell level function with a redundant `plam` - so `pelimList` would benefit greatly from just taking a Haskell level function directly.
```

Comment:
Consider:

It turns out that using `pelimList` *almost always* involves using a one-off Haskell level function (and therefore a redundant `plam`). In such cases, `pelimList` benefits greatly from just taking a Haskell level function directly.